wsbroadcastserver/clientconnection.go

```diff
@@ -76,49 +101,167 @@ func (cc *ClientConnection) Compression() bool {
+			requestedIdx := int(cc.requestedSeqNum) - int(prevSegment.Start())
+			msgs = msgs[requestedIdx:]
+		}
+		bm := &m.BroadcastMessage{
+			Version:  1, // TODO(clamb): I am unsure if it is correct to hard code the version here like this? It seems to be done in other places though
+			Messages: msgs,
+		}
+		err := cc.writeBroadcastMessage(bm)
+		if err != nil {
+			return err
+		}
+
+		// do not use prevSegment.End() method, must figure out the last
+		// sequence number from the messages that were actually sent in case
+		// more messages are added.
+		end := uint64(msgs[len(msgs)-1].SequenceNumber)
+		cc.LastSentSeqNum.Store(end)
+		log.Debug("segment sent to client", "client", cc.Name, "sentCount", len(bm.Messages), "lastSentSeqNum", end)
+	}
+	return nil
+}
+
+func (cc *ClientConnection) writeBroadcastMessage(bm *m.BroadcastMessage) error {
+	notCompressed, compressed, err := serializeMessage(bm, !cc.compression, cc.compression)
+	if err != nil {
+		return err
+	}
+
+	var data []byte
+	if cc.compression {
+		data = compressed.Bytes()
+	} else {
+		data = notCompressed.Bytes()
+	}
+	err = cc.writeRaw(data)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
 func (cc *ClientConnection) Start(parentCtx context.Context) {
 	cc.StopWaiter.Start(parentCtx, cc)
 	cc.LaunchThread(func(ctx context.Context) {
+		// A delay may be configured, ensures the Broadcaster delays before any
+		// messages are sent to the client. The ClientConnection has not been
+		// registered so the out channel filling is not a concern.
 		if cc.delay != 0 {
-			var delayQueue [][]byte
 			t := time.NewTimer(cc.delay)
-			done := false
-			for !done {
-				select {
-				case <-ctx.Done():
-					return
-				case data := <-cc.out:
-					delayQueue = append(delayQueue, data)
-				case <-t.C:
-					for _, data := range delayQueue {
-						err := cc.writeRaw(data)
-						if err != nil {
-							logWarn(err, "error writing data to client")
-							cc.clientManager.Remove(cc)
-							return
-						}
-					}
-					done = true
-				}
+			select {
+			case <-ctx.Done():
+				return
+			case <-t.C:
 			}
 		}
 
+		// Send the current backlog before registering the ClientConnection in
+		// case the backlog is very large
+		segment := cc.backlog.Head()
+		if !backlog.IsBacklogSegmentNil(segment) && segment.Start() < uint64(cc.requestedSeqNum) {
+			s, err := cc.backlog.Lookup(uint64(cc.requestedSeqNum))
+			if err != nil {
+				logWarn(err, "error finding requested sequence number in backlog: sending the entire backlog instead")
+			} else {
+				segment = s
+			}
+		}
+		err := cc.writeBacklog(ctx, segment)
+		if errors.Is(err, errContextDone) {
+			return
+		} else if err != nil {
+			logWarn(err, "error writing messages from backlog")
+			cc.Remove()
+			return
+		}
+
+		cc.Register()
+		timer := time.NewTimer(5 * time.Second)
+		select {
+		case <-ctx.Done():
+			return
+		case <-cc.registered:
+			log.Debug("ClientConnection registered with ClientManager", "client", cc.Name)
```

Comment:
I think we can remove this whole `registered` thing. Right now it seems to send a message to its own channel, and then pull it back out of the channel.

Reply:
Nevermind, I misread `register` and `registered`. I see how it works now.